codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res

```diff
@@ -0,0 +1,325 @@
+        | ("id", _)
+        | (_, Null(_)) => schema //TODO double check this works for array types
+        | _ => S.null(schema)->S.toUnknown
+        }
+
+        nulldict->Js.Dict.set(location, s.matches(nullableFieldSchema))
+      })
+      nulldict
+    | _ =>
+      Js.Exn.raiseError(
+        "Failed creating nullableEntitySchema. Expected an object schema for entity",
+      )
+    }
+  )
+
+  let previousWithNullFields = {
+    chain_id: None,
+    block_timestamp: None,
+    block_number: None,
+    log_index: None,
+  }
+
+  S.object(s => {
+    {
+      "current": s.flatten(currentHistoryFieldsSchema),
+      "previous": s.flatten(previousHistoryFieldsSchema),
+      "entityData": s.flatten(nullableEntitySchema),
+      "action": s.field("action", Enums.EntityHistoryRowAction.enum.schema),
+    }
+  })->S.transform(s => {
+    parser: v => {
+      current: v["current"],
+      previous: switch v["previous"] {
+      | {
+          chain_id: Some(chain_id),
+          block_timestamp: Some(block_timestamp),
+          block_number: Some(block_number),
+          log_index: Some(log_index),
+        } =>
+        Some({
+          chain_id,
+          block_timestamp,
+          block_number,
+          log_index,
+        })
+      | {chain_id: None, block_timestamp: None, block_number: None, log_index: None} => None
+      | _ => s.fail("Unexpected mix of null and non-null values in previous history fields")
+      },
+      entityData: switch v["action"] {
+      | SET => v["entityData"]->S.parseAnyOrRaiseWith(entitySchema)->Set
+      | DELETE => v["entityData"]->S.parseAnyOrRaiseWith(entityIdOnlySchema)->Delete
+      },
+    },
+    serializer: v => {
+      let (entityData, action) = switch v.entityData {
+      | Set(entityData) => (
+          entityData->(Utils.magic: 'entity => Js.Dict.t<unknown>),
+          Enums.EntityHistoryRowAction.SET,
+        )
+      | Delete({id}) => (getEntityWithNullFields(id), DELETE)
+      }
+
+      {
+        "current": v.current,
+        "entityData": entityData,
+        "action": action,
+        "previous": switch v.previous {
+        | Some(historyFields) =>
+          historyFields->(Utils.magic: historyFields => previousHistoryFields) //Cast to previousHistoryFields (with "Some" field values)
+        | None => previousWithNullFields
+        },
+      }
+    },
+  })
+}
+
+type t<'entity> = {
+  table: table,
+  createInsertFnQuery: string,
+  schema: S.t<historyRow<'entity>>,
+  insertFn: (Postgres.sql, Js.Json.t) => promise<unit>,
+}
+
+let insertRow = (self: t<'entity>, ~sql, ~historyRow: historyRow<'entity>) => {
+  let row = historyRow->S.serializeOrRaiseWith(self.schema)
+  self.insertFn(sql, row)
+}
+
+let batchInsertRows = (self: t<'entity>, ~sql, ~rows: array<historyRow<'entity>>) => {
+  Utils.Array.awaitEach(rows, async row => {
+    let row = row->S.serializeOrRaiseWith(self.schema)
+    await self.insertFn(sql, row)
```

Comment:
Should we maybe have an insertFn which works with multiple items?

Reply:
I assume it's going to be faster

Reply:
As in - a postgres function that iterates through a set of rows and calls the postgres insert function?

We can try that, might be a bit faster 👍🏼

Think I will just try get everything end to end first and note this as an optimisation possibility.